include/deal.II/lac/vector_operations_internal.h

```diff
@@ -210,14 +210,9 @@ namespace internal
         Assert(end >= begin, ExcInternalError());
 
         if (value == Number())
-          {
-            if constexpr (std::is_trivial_v<Number>)
-              {
-                std::memset(dst + begin, 0, sizeof(Number) * (end - begin));
-                return;
-              }
-          }
-        std::fill(dst + begin, dst + end, value);
+          std::fill(dst + begin, dst + end, Number());
+        else
+          std::fill(dst + begin, dst + end, value);
```

Comment:
This was also hard to read, with the entirely avoidable `return` in the middle. The new code still differentiates between setting to zero and something else, and allows the compiler to optimize based on it.